hacks/glx/kallisti.man
.TH XScreenSaver 1 "" "X Version 11"
.SH NAME
kallisti \- the golden apple of discord.
.SH SYNOPSIS
.B kallisti
[\-\-display \fIhost:display.screen\fP]
[\-\-visual \fIvisual\fP]
[\-\-window]
[\-\-root]
[\-\-delay \fInumber\fP]
[\-\-speed \fInumber\fP]
[\-\-wander]
[\-\-spin 0]
[\-\-spin \fI[XYZ]\fP]
[\-\-spin Z]
[\-\-spin XY]
[\-\-spin XZ]
[\-\-spin YZ]
[\-\-spin XYZ]
[\-\-fps]
.SH DESCRIPTION
The golden apple of discord.

Greek goddess Eris was not invited to a wedding; her retort was to toss in a
golden apple, inscribed: "to the fairest". Hera, Athena and Aphrodite
squabbled over it; Zeus noped out, telling Paris of Troy to decide. They all
cheated, but Aphrodite won by letting Paris abduct and marry Helen of
Sparta. The Greeks objected to this, leading to the Trojan War.

Do you want to launch a thousand ships?  Because this is how you launch a
thousand ships.
.SH OPTIONS
.TP 8
.B \-\-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-\-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-\-root
Draw on the root window.
.TP 8
.B \-\-delay \fInumber\fP
Per-frame delay, in microseconds.  Default: 30000 (0.03 seconds).
.TP 8
.B \-\-speed \fInumber\fP
Animation speed.  2.0 means twice as fast, 0.5 means half as fast.
.TP 8
.B \-\-wander | \-\-no\-wander
Whether the object should wander around the screen.
.TP 8
.B \-\-spin 0
Don't rotate.
.TP 8
.B \-\-spin \fI[XYZ]\fP
Around which axes should the object spin?
.TP 8
.B \-\-spin X
Rotate around X axis.
.TP 8
.B \-\-spin Y
Rotate around Y axis. Default.
.TP 8
.B \-\-spin Z
Rotate around Z axis.
.TP 8
.B \-\-spin XY
Rotate around X and Y axes.
.TP 8
.B \-\-spin XZ
Rotate around X and Z axes.
.TP 8
.B \-\-spin YZ
Rotate around Y and Z axes.
.TP 8
.B \-\-spin XYZ
Rotate around all three axes.
.TP 8
.B \-\-fps | \-\-no\-fps
Whether to show a frames-per-second display at the bottom of the screen.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 8
.B XSCREENSAVER_WINDOW
The window ID to use with \fI\-\-root\fP.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1)
.SH COPYRIGHT
Copyright \(co 2023 by Jamie Zawinski and Jared Williams.  Permission to use,
copy, modify, distribute, and sell this software and its documentation for any
purpose is hereby granted without fee, provided that the above copyright
notice appear in all copies and that both that copyright notice and this
permission notice appear in supporting documentation.  No representations are
made about the suitability of this software for any purpose.  It is provided
"as is" without express or implied warranty.
.SH AUTHOR
Code by Jamie Zawinski; model by Jared Williams.


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/glx/kallisti.c
/* kallisti, Copyright © 2023-2024 Jamie Zawinski <jwz@jwz.org>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 *
 * The golden apple of discord: https://www.jwz.org/blog/2023/09/ti-kallisti/
 */

#define DEFAULTS	"*delay:	30000       \n" \
			"*showFPS:      False       \n" \
			"*wireframe:    False       \n" \

# define release_chao 0
#define DEF_SPIN        "Y"
#define DEF_WANDER      "FALSE"
#define DEF_SPEED       "1.0"

#include "xlockmore.h"
#include "rotator.h"
#include "gltrackball.h"
#include <ctype.h>

#ifdef USE_GL /* whole file */

#include "gllist.h"

extern struct gllist *kallisti_model;


typedef struct {
  GLXContext *glx_context;
  trackball_state *trackball;
  rotator *rot;
  Bool button_down_p;
  GLuint chao_list;
  GLuint texture;
  enum { FADE_IN, DRAW } state;
  GLfloat tick;
  GLfloat th;
} chao_configuration;

static chao_configuration *bps = NULL;

static char *do_spin;
static GLfloat speed;
static Bool do_wander;

static XrmOptionDescRec opts[] = {
  { "-spin",   ".spin",   XrmoptionSepArg, 0 },
  { "+spin",   ".spin",   XrmoptionNoArg, "" },
  { "-speed",  ".speed",  XrmoptionSepArg, 0 },
  { "-wander", ".wander", XrmoptionNoArg, "True" },
  { "+wander", ".wander", XrmoptionNoArg, "False" }
};

static argtype vars[] = {
  {&do_spin,   "spin",   "Spin",   DEF_SPIN,   t_String},
  {&do_wander, "wander", "Wander", DEF_WANDER, t_Bool},
  {&speed,     "speed",  "Speed",  DEF_SPEED,  t_Float},
};

ENTRYPOINT ModeSpecOpt chao_opts = {
  countof(opts), opts, countof(vars), vars, NULL};


ENTRYPOINT void
reshape_chao (ModeInfo *mi, int width, int height)
{
  GLfloat h = (GLfloat) height / (GLfloat) width;
  int y = 0;

  if (width > height * 5) {   /* tiny window: show middle */
    height = width * 9/16;
    y = -height/2;
    h = height / (GLfloat) width;
  }

  glViewport (0, y, (GLint) width, (GLint) height);

  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  gluPerspective (30.0, 1/h, 1.0, 100);

  glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
  gluLookAt( 0.0, 0.0, 30.0,
             0.0, 0.0, 0.0,
             0.0, 1.0, 0.0);

  {
    GLfloat s = (MI_WIDTH(mi) < MI_HEIGHT(mi)
                 ? (MI_WIDTH(mi) / (GLfloat) MI_HEIGHT(mi))
                 : 1);
    glScalef (s, s, s);
  }

  glClear(GL_COLOR_BUFFER_BIT);
}


ENTRYPOINT Bool
chao_handle_event (ModeInfo *mi, XEvent *event)
{
  chao_configuration *bp = &bps[MI_SCREEN(mi)];

  if (gltrackball_event_handler (event, bp->trackball,
                                 MI_WIDTH (mi), MI_HEIGHT (mi),
                                 &bp->button_down_p))
    return True;

  return False;
}


ENTRYPOINT void
init_chao (ModeInfo *mi)
{
  chao_configuration *bp;
  int wire = MI_IS_WIREFRAME(mi);

  MI_INIT (mi, bps);
  bp = &bps[MI_SCREEN(mi)];

  bp->glx_context = init_GL(mi);

  reshape_chao (mi, MI_WIDTH(mi), MI_HEIGHT(mi));

  {
    Bool spinx=False, spiny=False, spinz=False;
    double spin_speed   = 0.6 * speed;
    double spin_accel   = 0.3;
    double wander_speed = 0.01 * speed;

    char *s = do_spin;
    while (*s)
      {
        if      (*s == 'x' || *s == 'X') spinx = True;
        else if (*s == 'y' || *s == 'Y') spiny = True;
        else if (*s == 'z' || *s == 'Z') spinz = True;
        else if (*s == '0') ;
        else
          {
            fprintf (stderr,
         "%s: spin must contain only the characters X, Y, or Z (not \"%s\")\n",
                     progname, do_spin);
            exit (1);
          }
        s++;
      }

    bp->rot = make_rotator (spinx ? spin_speed : 0,
                            spiny ? spin_speed : 0,
                            spinz ? spin_speed : 0,
                            spin_accel,
                            do_wander ? wander_speed : 0,
                            False);
    bp->trackball = gltrackball_init (True);
  }

  bp->chao_list = glGenLists (1);
  glNewList (bp->chao_list, GL_COMPILE);
  renderList (kallisti_model, wire);
  glEndList ();

  bp->state = FADE_IN;
  bp->tick = 0;
}


ENTRYPOINT void
draw_chao (ModeInfo *mi)
{
  chao_configuration *bp = &bps[MI_SCREEN(mi)];
  Display *dpy = MI_DISPLAY(mi);
  Window window = MI_WINDOW(mi);

  if (!bp->glx_context)
    return;

  glXMakeCurrent(MI_DISPLAY(mi), MI_WINDOW(mi), *bp->glx_context);

  glShadeModel(GL_SMOOTH);

  glEnable(GL_DEPTH_TEST);
  glEnable(GL_NORMALIZE);
  glEnable(GL_CULL_FACE);

  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

  switch (bp->state) {
  case FADE_IN:
    bp->tick += 0.01 * speed;
    if (bp->tick >= 1.0)
      {
        bp->tick = 1;
        bp->state = DRAW;
      }
    break;
  case DRAW:
    break;
  default:
    abort();
  }

  if (! MI_IS_WIREFRAME(mi))
    {
      GLfloat pos0[]  = { -30.0,  15.0,   3.0, 1.0 };
      GLfloat pos1[]  = {  24.0, -12.0, -12.0, 1.0 };
      GLfloat pos2[]  = {   0.0,   0.0,   0.0, 1.0 };

      GLfloat amb0[]  = { 0.4, 0.4, 0.4, 1.0 };
      GLfloat amb1[]  = { 0.0, 0.0, 0.0, 1.0 };
      GLfloat amb2[]  = { 0.0, 0.0, 0.0, 1.0 };
      GLfloat dif0[]  = { 1.0, 1.0, 1.0, 1.0 };
      GLfloat dif1[]  = { 0.7, 0.1, 0.1, 1.0 };
      GLfloat dif2[]  = { 0.6, 0.6, 0.0, 1.0 };

      amb0[0] *= bp->tick; amb0[1] *= bp->tick; amb0[2] *= bp->tick;
      dif0[0] *= bp->tick; dif0[1] *= bp->tick; dif0[2] *= bp->tick;
      amb1[0] *= bp->tick; amb1[1] *= bp->tick; amb1[2] *= bp->tick;
      dif1[0] *= bp->tick; dif1[1] *= bp->tick; dif1[2] *= bp->tick;
      amb2[0] *= bp->tick; amb2[1] *= bp->tick; amb2[2] *= bp->tick;
      dif2[0] *= bp->tick; dif2[1] *= bp->tick; dif2[2] *= bp->tick;

      glLightfv(GL_LIGHT0, GL_POSITION, pos0);
      glLightfv(GL_LIGHT1, GL_POSITION, pos1);

      glLightfv(GL_LIGHT0, GL_AMBIENT,  amb0);
      glLightfv(GL_LIGHT1, GL_AMBIENT,  amb1);
      glLightfv(GL_LIGHT2, GL_AMBIENT,  amb2);
      glLightfv(GL_LIGHT0, GL_DIFFUSE,  dif0);
      glLightfv(GL_LIGHT1, GL_DIFFUSE,  dif1);
      glLightfv(GL_LIGHT2, GL_DIFFUSE,  dif2);

      glEnable(GL_LIGHTING);
      glEnable(GL_LIGHT0);
      glEnable(GL_LIGHT1);
      glEnable(GL_LIGHT2);

      glEnable(GL_DEPTH_TEST);
      glEnable(GL_TEXTURE_2D);
      glEnable(GL_NORMALIZE);
      glEnable(GL_CULL_FACE);

      { /* Spinny light */
        GLfloat s = 10;
        GLfloat th2 = M_PI * 0.65;

        bp->th += 0.02 * speed;
        while (bp->th > M_PI * 2)
          bp->th -= M_PI * 2;

        pos2[0] = s * cos (th2) * cos (-bp->th);
        pos2[1] = s * sin (-bp->th);
        pos2[2] = s * sin (th2) * cos (-bp->th);
        glLightfv (GL_LIGHT2, GL_POSITION, pos2);

# if 0
        glDisable (GL_LIGHTING);
        glBegin (GL_LINES);
        glVertex3f (0, 0, 0);
        glVertex3fv (pos2);
        glEnd();
        glEnable (GL_LIGHTING);
# endif
      }

    }

  glPushMatrix ();
  glRotatef (current_device_rotation(), 0, 0, 1);
  glScalef (10, 10, 10);

  glRotatef (15, 1, 0, 0);
  glTranslatef (0, -0.3, 0);

  {
    double x, y, z;
    get_position (bp->rot, &x, &y, &z, !bp->button_down_p);
    glTranslatef((x - 0.5) * 1,
                 (y - 0.5) * 0.8,
                 (z - 1) * 2);

    gltrackball_rotate (bp->trackball);

    get_rotation (bp->rot, &x, &y, &z, !bp->button_down_p);
    glRotatef (x * 360, 1.0, 0.0, 0.0);
    glRotatef (y * 360, 0.0, 1.0, 0.0);
    glRotatef (z * 360, 0.0, 0.0, 1.0);
  }

  glRotatef (-90, 1, 0, 0);


  mi->polygon_count = 0;

  {
    GLfloat amb[] = { 0.33, 0.22, 0.03, 1.00 };
    GLfloat spc[] = { 0.78, 0.57, 0.11, 1.00 };
    GLfloat dif[] = { 0.99, 0.91, 0.81, 1.00 };
    GLfloat shiny = 27.80;

    amb[0] *= bp->tick; amb[1] *= bp->tick; amb[2] *= bp->tick;
    spc[0] *= bp->tick; spc[1] *= bp->tick; spc[2] *= bp->tick;
    dif[0] *= bp->tick; dif[1] *= bp->tick; dif[2] *= bp->tick;

    glMaterialfv(GL_FRONT, GL_AMBIENT,   amb);
    glMaterialfv(GL_FRONT, GL_DIFFUSE,   spc);
    glMaterialfv(GL_FRONT, GL_SPECULAR,  dif);
    glMaterialf (GL_FRONT, GL_SHININESS, shiny);
  }

# if 0
  glDisable (GL_LIGHTING);
  glBegin (GL_LINES);
  glVertex3f (0, 0, -100); glVertex3f (0, 0, 100);
  glVertex3f (0, -100, 0); glVertex3f (0, 100, 0);
  glVertex3f (-100, 0, 0); glVertex3f (100, 0, 0);
  glEnd();
  glEnable (GL_LIGHTING);
# endif

  /* Align axis with core of apple, since the model's bbox isn't centered. */
  glTranslatef (0.0534, 0.0394, -0.03);

  glCallList (bp->chao_list);
  mi->polygon_count += kallisti_model->points / 3;

  glPopMatrix ();

  if (mi->fps_p) do_fps (mi);
  glFinish();

  glXSwapBuffers(dpy, window);
}


ENTRYPOINT void
free_chao (ModeInfo *mi)
{
  chao_configuration *bp = &bps[MI_SCREEN(mi)];
  if (!bp->glx_context) return;
  glXMakeCurrent(MI_DISPLAY(mi), MI_WINDOW(mi), *bp->glx_context);
  free_rotator (bp->rot);
  if (glIsList(bp->chao_list)) glDeleteLists(bp->chao_list, 1);
  if (bp->trackball) gltrackball_free (bp->trackball);
}

XSCREENSAVER_MODULE_2 ("Kallisti", kallisti, chao)

#endif /* USE_GL */
